Formula/r/redress.rb

```diff
@@ -0,0 +1,40 @@
+class Redress < Formula
+  desc "Tool for analyzing stripped Go binaries compiled with the Go compiler"
+  homepage "https://go-re.tk/redress/"
+  url "https://github.com/goretk/redress/archive/refs/tags/v1.1.1.tar.gz"
+  sha256 "5c938791f2010233054ede11401ed7a00e32151f37042f95e08d103e46d98f53"
+  license "AGPL-3.0-only"
+  head "https://github.com/goretk/redress.git", branch: "develop"
+
+  livecheck do
+    url :stable
+    regex(/^v?(\d+(?:\.\d+)+)$/i)
+  end
+
+  depends_on "go" => :build
+
+  def install
+    # https://github.com/goretk/redress/blob/develop/Makefile#L11-L14
+    redress_version = version
+    gore_version = File.read(buildpath/"go.mod").scan(%r{goretk/gore v(\S+)}).flatten.first
+    go_version = Formula["go"].version
+
+    ldflags = %W[
+      -s -w
+      -X main.redressVersion=#{redress_version}
+      -X main.goreVersion=#{gore_version}
+      -X main.compilerVersion=#{go_version}
+    ]
+
+    system "go", "build", *std_go_args(ldflags: ldflags)
+
+    generate_completions_from_executable(bin/"redress", "completion")
+  end
+
+  test do
+    assert_match "Version:  #{version}", shell_output("#{bin}/redress version")
+
+    output = shell_output("#{bin}/redress info '#{bin}/redress'")
```

Comment:
```
$ /opt/homebrew/Cellar/redress/1.1.1/bin/redress info /opt/homebrew/Cellar/redress/1.1.1/bin/redress
panic: Unsupported architecture

goroutine 1 [running]:
github.com/goretk/gore.(*machoFile).getFileInfo(0x1400002c4a0)
	github.com/goretk/gore@v0.11.1/macho.go:117 +0xdc
github.com/goretk/gore.Open({0x16f1e6a84, 0x2e})
	github.com/goretk/gore@v0.11.1/file.go:86 +0x348
main.listInfo({0x16f1e6a84?, 0x0?})
	github.com/goretk/redress/info.go:50 +0x90
main.init.1.func1(0x140000e6e00?, {0x1400002c300?, 0x4?, 0x100e4f6df?})
	github.com/goretk/redress/info.go:25 +0x30
github.com/spf13/cobra.(*Command).execute(0x140000c4280, {0x1400002c2e0, 0x1, 0x1})
	github.com/spf13/cobra@v1.2.1/command.go:860 +0x53c
github.com/spf13/cobra.(*Command).ExecuteC(0x1010dd2c0)
	github.com/spf13/cobra@v1.2.1/command.go:974 +0x318
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.2.1/command.go:902
main.main()
	github.com/goretk/redress/main.go:25 +0x24
```

is this only for x86?

Reply:
That seems to be coming from the `gore` lib that `redress` depends on:

- https://github.com/goretk/gore/blob/78568206ae24043a24fc8bf89443ad6164a7509f/macho.go#L104-L123

```go
func (m *machoFile) getFileInfo() *FileInfo {
	fi := &FileInfo{
		ByteOrder: m.file.ByteOrder,
		OS:        "macOS",
	}
	switch m.file.Cpu {
	case macho.Cpu386:
		fi.WordSize = intSize32
		fi.Arch = Arch386
	case macho.CpuAmd64:
		fi.WordSize = intSize64
		fi.Arch = ArchAMD64
	case macho.CpuArm64:
		fi.WordSize = intSize64
		fi.Arch = ArchARM64
	default:
		panic("Unsupported architecture")
	}
	return fi
}
```

---

> is this only for x86?

@chenrui333 Does the above answer your question? I'm not sure how to otherwise check/test/answer.

Reply:
I am trying on ventura arm, if does not run for arm, then we probably need `depends_on arch: :x86_64` as well as some upstream note on it

Reply:
Also opened an issue against the `redress` repo referencing that other dependency issue for better findability:

- https://github.com/goretk/redress/issues/35

Reply:
> (unverified) potentially could cross-compile a small golang program and force it to be non-ARM (even when running on an ARM system), and then process that

Even without testing my theory on an ARM system, I'm just going to continue with this path, as I think it's going to be the correct way forward regardless.

For my reference, here are other formula that seem to depend on Go for their `:test`s, to use as a reference:

- https://github.com/search?q=repo%3A0xdevalias%2Fhomebrew-core%20depends_on%20%5C%22go%5C%22%20%3D%3E%20%3Abuild%2C%20%3Atest&type=code

Reply:
Only just noticed that these test errors were annotated on the file view just now (was previously viewing your review comments from the main conversation view):

> `brew test --verbose redress` failed on macOS Sonoma (14) on Apple Silicon!
> `brew test --verbose redress` failed on macOS Ventura (13) on Apple Silicon!
> `brew test --verbose redress` failed on macOS Monterey (12) on Apple Silicon!

Reply:
Looks like we can conditionally change the test code based on the system:

- https://docs.brew.sh/Formula-Cookbook#inside-def-install-and-test-do
  - > Inside def install and test do, don’t use these `on_*` methods. Instead, use `if` statements and the following conditionals:
  - > `OS.mac?` and `OS.linux?` return `true` or `false` based on the OS
`Hardware::CPU.intel?` and `Hardware::CPU.arm?` return `true` or `false` based on the arch

Reply:
Ok, got some better test code working locally, and as a sanity check, used it to cross-compile an apple silicon (`darwin-arm64`) golang binary, which fails in the same way as we're seeing on the CI:

```shell
 brew test redress
==> Testing redress
==> /usr/local/Cellar/redress/1.1.1/bin/redress version
==> go mod init brewtest
==> /usr/local/opt/go/bin/go build -o /private/tmp/redress-test-20231129-86693-1xh0btf/brewtest-darwin-arm64 -ldflags -s -w /private/tmp/redress-test-20231129-86693-1xh0btf/main.go
==> /usr/local/Cellar/redress/1.1.1/bin/redress info '/private/tmp/redress-test-20231129-86693-1xh0btf/brewtest-darwin-arm64'
panic: Unsupported architecture

goroutine 1 [running]:
github.com/goretk/gore.(*machoFile).getFileInfo(0xc0000a8460)
	github.com/goretk/gore@v0.11.1/macho.go:117 +0xc5
github.com/goretk/gore.Open({0x7ff7bfefe141, 0x46})
	github.com/goretk/gore@v0.11.1/file.go:86 +0x324
main.listInfo({0x7ff7bfefe141?, 0x0?})
	github.com/goretk/redress/info.go:50 +0xb7
main.init.1.func1(0xc00011ce00?, {0xc0000a82c0?, 0x4?, 0x12da053?})
	github.com/goretk/redress/info.go:25 +0x25
github.com/spf13/cobra.(*Command).execute(0xc0000fc280, {0xc0000a82a0, 0x1, 0x1})
	github.com/spf13/cobra@v1.2.1/command.go:860 +0x67b
github.com/spf13/cobra.(*Command).ExecuteC(0x14ed2e0)
	github.com/spf13/cobra@v1.2.1/command.go:974 +0x38d
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.2.1/command.go:902
main.main()
	github.com/goretk/redress/main.go:25 +0x1a
```

Which means that my theory as to why it was failing seems correct, and the new test code should resolve the issue without having to resort to using something like `depends_on arch: :x86_64`, as I am almost certain the redress binary will fine on both architectures.

Reply:
that sounds good 👍

Reply:
Just pushed my new fixed test block; all seems to work fine locally, so let's hope the CI agrees:

- https://github.com/Homebrew/homebrew-core/pull/155724/commits/15341de85a0438c0a15e8b897ee021325f63e8da

Reply:
🤞

Reply:
> `brew test --verbose redress` failed on Linux!

I know why this is (`GOOS` and `GOARCH` env vars aren't always set by default, so we need to get their defaults since I use the values in my assertions, which are currently blank); and have a fix in-progress.

---

@chenrui333 Formula style question.. which is a 'better'/more canonical choice out of the following, and if possible, why?

My personal preference for simplicity/consistency with the other methods in the `test` block would be 'Choice C'; but I'm not sure if there are any issues with that, and whether these other formulae do it differently for a specific reason, or just because legacy reasons.

**Choice A:**

```ruby
goos = Utils.safe_popen_read("#{Formula["go"].bin}/go", "env", "GOOS").chomp
goarch = Utils.safe_popen_read("#{Formula["go"].bin}/go", "env", "GOARCH").chomp
```

**Choice B:**

```ruby
goos = shell_output("#{Formula["go"].bin}/go env GOOS").strip
goarch = shell_output("#{Formula["go"].bin}/go env GOARCH").strip
```

**Choice C:**

```ruby
goos = shell_output("go env GOOS").strip
goarch = shell_output("go env GOARCH").strip
```

In the cookbook, I see `shell_output` mentioned, but not `safe_popen_read`:

- https://docs.brew.sh/Formula-Cookbook

Yet some of the existing formula examples I'm looking at are using `Utils.safe_popen_read` for this sort of task, eg.:

- https://github.com/Homebrew/homebrew-core/blob/master/Formula/b/brigade-cli.rb#L36-L37
  - (though it's using them in `install`)
- https://github.com/Homebrew/homebrew-core/blob/master/Formula/g/garble.rb#L41-L42
  - (using them in `test`)
- https://github.com/Homebrew/homebrew-core/blob/master/Formula/k/kubebuilder.rb#L23-L24
  - (though it's using them in `install`)

Looking at where `shell_output` is defined, it's part of `Homebrew::Assertions`, which makes me think maybe it shouldn't be used outside of the `test` block (which I'm not doing here):

- https://rubydoc.brew.sh/Homebrew/Assertions.html#shell_output-instance_method
  - > Returns the output of running cmd, and asserts the exit status.
  - Looking at the code, it uses backticks to run the command, and the does an `assert_equal result, $CHILD_STATUS.exitstatus` before returning the `output` (as well as logging the output to the console `if verbose?`

Whereas `safe_popen_read` is defined in `Utils`:

- https://rubydoc.brew.sh/Utils.html#safe_popen_read-class_method
  - `safe_popen_read` calls `popen_read` with `safe: true`
- https://rubydoc.brew.sh/Utils.html#popen_read-class_method
  - `popen_read` calls `popen`, but only returns the output if `output if !safe || $CHILD_STATUS.success?`
  - it also seems to be able to raise an `ErrorDuringExecution`
- https://rubydoc.brew.sh/Utils.html#popen-class_method
  - `popen` uses `IO.popen` to create a `pipe`, and then calls `exec`, with some `rescue`'s for `Errno::ENOENT` and `SystemCallError`

Reply:
you are correct, I would just go with choice A, I never saw choice b/c in install block.

Reply:
I will take it back, if just for test, both b or c are kind of equivalent.

Reply:
> I never saw choice b/c in install block

@chenrui333 This isn't being used in `install` block, but in `test` block

> I would just go with choice A

@chenrui333 Hrmm.. fair. I'd already implemented it with choice C here:

- https://github.com/Homebrew/homebrew-core/pull/155724/commits/224267c064fc724850e2df8cfa303f28377fd9c8#diff-9bcc514770fedd71be3c57dcd07d9f37eca5993f280cc2714c16979c37168af1R63-R65

---

It would be nice if `shell_output` called `Utils.safe_popen_read` under the hood so that there aren't disparate ways of doing things. Is there any specific reason it's not implemented that way?

Reply:
> `brew test --verbose redress` failed on Linux!
> 
> An exception occurred within a child process:
>   Minitest::Assertion: Expected /GOARCH\s+amd64/ to match "OS         EM_X86_64\nArch       amd64\nCompiler   1.4 (2014-12-11)\nBuild ID   eM9GLfPXXJOAQtiwTT6m/Su8G1c67K6CrsETe-S4p/7EhXh2RSzSnISnA3JOm6/WcR9hESpy2sgco4tYxlc\nMain root  github.com/goretk/redress\n# main     1\n# std      74\n# vendor   7\n# unknown  2\n".

It seems the `GOOS` and `GOARCH` aren't shown in the output from `redress info` for the linux build anyway, which is apparently what was breaking this test.

So in the end.. the above is all kind of moot it seems.

Am just going to remove the 2 extra asserts I was doing, as the one on for 'Main root' is technically enough to prove the binary is running.

Reply:
Ok, fingers crossed this will be the final time:

- https://github.com/Homebrew/homebrew-core/pull/155724/commits/b6d8d6592a9b91628cf6634e23119c5970d8943f

Reply:
Omg.. it actually passed finally! 🎉